community/kernel/src/main/java/org/neo4j/kernel/api/schema/MultiTokenSchemaDescriptor.java

```diff
@@ -0,0 +1,167 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.schema;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import org.neo4j.internal.kernel.api.TokenNameLookup;
+import org.neo4j.internal.kernel.api.schema.SchemaComputer;
+import org.neo4j.internal.kernel.api.schema.SchemaDescriptor;
+import org.neo4j.internal.kernel.api.schema.SchemaProcessor;
+import org.neo4j.internal.kernel.api.schema.SchemaUtil;
+import org.neo4j.kernel.impl.locking.ResourceTypes;
+import org.neo4j.storageengine.api.EntityType;
+import org.neo4j.storageengine.api.lock.ResourceType;
+
+import static java.util.Arrays.binarySearch;
+import static java.util.stream.Collectors.joining;
+
+public class MultiTokenSchemaDescriptor implements org.neo4j.internal.kernel.api.schema.MultiTokenSchemaDescriptor
+{
+    private final int[] entityTokens;
+    private final EntityType entityType;
+    private final int[] propertyIds;
+
+    MultiTokenSchemaDescriptor( int[] entityTokens, EntityType entityType, int[] propertyIds )
+    {
+        this.entityTokens = entityTokens;
+        Arrays.sort( this.entityTokens );
+        this.entityType = entityType;
+        this.propertyIds = propertyIds;
+    }
+
+    @Override
+    public boolean isAffected( long[] entityIds )
+    {
+        if ( entityTokens.length == 0 )
+        {
+            return true;
+        }
+        return Arrays.stream( entityTokens ).anyMatch( id -> ArrayUtils.contains( entityIds, id ) );
+    }
+
+    @Override
+    public <R> R computeWith( SchemaComputer<R> computer )
+    {
+        return computer.computeSpecific( this );
+    }
+
+    @Override
+    public void processWith( SchemaProcessor processor )
+    {
+        processor.processSpecific( this );
+    }
+
+    @Override
+    public String toString()
+    {
+        return "MultiTokenSchemaDescriptor[" + userDescription( SchemaUtil.idTokenNameLookup ) + "]";
+    }
+
+    @Override
+    public String userDescription( TokenNameLookup tokenNameLookup )
+    {
+        return String.format( entityType + ":%s(%s)",
+                Arrays.stream( tokenNameLookup.entityTokensGetNames( entityType, entityTokens ) )
+                      .collect( joining( ", " ) ),
+                SchemaUtil.niceProperties( tokenNameLookup, propertyIds ) );
+    }
+
+    @Override
+    public int[] getPropertyIds()
+    {
+        return propertyIds;
+    }
+
+    @Override
+    public int[] getEntityTokenIds()
+    {
+        return entityTokens;
+    }
+
+    @Override
+    public int keyId()
+    {
+        //TODO MultiTokenSchema support for the new kernel api. This method is only used in the AllStoreHolder
```

Comment:
TODO

Reply:
This is the bleeding edge towards the not-yet-supported new kernel api.